local Dialog = require("@self/Dialog");
local Icon = require("@self/Icon");
local NavigationTab = require("@self/NavigationTab");
local NavigationBar = require("@self/NavigationBar");
local Ripple = require("@self/Ripple");
local TabItem = require("@self/TabItem");
local Tabs = require("@self/Tabs");
local FilledTextButton = require("@self/Button/FilledTextButton");
local FilledTonalTextButton = require("@self/Button/FilledTonalTextButton");
local TextButton = require("@self/Button/TextButton");
local Switch = require("@self/Switch");
local SideSheet = require("@self/SideSheet");
local IconButton = require("@self/IconButton");
local Menu = require("@self/Menu");
local Item = require("@self/Item");
local Shadow = require("@self/Shadow");
local AlertDialog = require("@self/AlertDialog");
local CircularProgressIndicator = require("@self/CircularProgress");
local Checkbox = require("@self/Checkbox");
local RadioButton = require("@self/RadioButton");
local TextField = require("@self/TextField");
local Card = require("@self/Card");
local Snackbar = require("@self/Snackbar");
local FloatingActionButton = require("@self/FloatingActionButton");
local Divider = require("@self/Divider");
local Chip = require("@self/Chip");
local LinearProgress = require("@self/LinearProgress");
local Badge = require("@self/Badge");
local Tooltip = require("@self/Tooltip");

return {
    Dialog = Dialog,
    Icon = Icon,
    NavigationTab = NavigationTab,
    NavigationBar = NavigationBar,
    Ripple = Ripple,
    TabItem = TabItem,
    Tabs = Tabs,
    FilledTextButton = FilledTextButton,
    FilledTonalTextButton = FilledTonalTextButton,
    TextButton = TextButton,
    Switch = Switch,
    SideSheet = SideSheet,
    IconButton = IconButton,
    Menu = Menu,
    Item = Item,
    Shadow = Shadow,
    AlertDialog = AlertDialog,
    CircularProgressIndicator = CircularProgressIndicator,
    Checkbox = Checkbox,
    RadioButton = RadioButton,
    TextField = TextField,
    Card = Card,
    Snackbar = Snackbar,
    FloatingActionButton = FloatingActionButton,
    Divider = Divider,
    Chip = Chip,
    LinearProgress = LinearProgress,
    Badge = Badge,
    Tooltip = Tooltip,
}